src/blog/internet-computer-web-app-decentralized-database-architecture.md

```diff
@@ -0,0 +1,345 @@
+
+*****
+
+#### Code Snippet
+
+We would need two actors to implement such a solution. One that acts as a “manager”, the canister that generate other canisters on the fly, and, an implementation for those that take care of the data persistence, those that are generated on the fly.
+
+The “manager” might have to track or not the list of canisters it has generated. If end users save on their side their references, such feature might not be needed. However, in following snippet we will assume it does has to keep track of what it generates.
+
+![](https://cdn-images-1.medium.com/max/1600/1*i1UedTMw4GxvoiwkQI9gvw.png)
+
+The base implementation of the manager is close to what we already implemented. However, instead of `Trie`, I used a [HashMap](https://sdk.dfinity.org/docs/base-libraries/hashmap) to keep track of the IDs of the canisters that has been generated.
+
+```
+import HashMap "mo:base/HashMap";
+import Principal "mo:base/Principal";
+import Cycles "mo:base/ExperimentalCycles";
+import Iter "mo:base/Iter";
+
+import Pet "./Pet";
+
+actor {
+
+  type CanisterId = Principal;
+  type UserId = Principal;
+
+  private func isPrincipalEqual(x: Principal, y: Principal): Bool {
+    x == y 
+  };
+
+  private var canisters: HashMap.HashMap<UserId, CanisterId> = 
+    HashMap.HashMap<UserId, CanisterId>(10, isPrincipalEqual,  
+                                        Principal.hash);
+
+  private stable var upgradeCanisters: [(Principal, CanisterId)] = 
+    [];
+
+  public shared({caller}) func create(): async (CanisterId) {
+    Cycles.add(1_000_000_000_000);
+    let canister = await Pet.Pet();
+    let id: CanisterId = await canister.id();
+    canisters.put(caller, id);
+
+    return id;
+  };
+
+  public shared query({caller}) func get() : async ?CanisterId {
+    let id: ?CanisterId = canisters.get(caller);
+    return id;
+  };
+
+  system func preupgrade() {
+    upgradeCanisters := Iter.toArray(canisters.entries());
+  };
+
+  system func postupgrade() {
+    canisters := HashMap.fromIter<UserId, CanisterId>     
+                  (upgradeCanisters.vals(), 10, 
+                  isPrincipalEqual, Principal.hash);
+                  upgradeCanisters := [];
+  };
+
+};
+```
+
+The function `create` takes care of unleashing new user’s canisters in the Internet Computer, those that will be used for the data persistence. It allocates first some few minimal cycles for the newly actors.
+
+The `get` function returns, if it exists, the ID of the user’s canister.
+
+Both `preupgrade` and `postupgrade` are used to persist the memory among upgrades (see [documentation](https://sdk.dfinity.org/docs/language-guide/upgrades.html)).
+
+Once the manager being set, we would be able to implement an `actor` dedicated to the personal data persistence of a user.
```

Comment:
```suggestion
Once the manager being set, we would be able to implement an `actor` dedicated to the personal data persistence of a user.
```

I don't understand what you mean bu "once the manager is set"?

Reply:
I'll replace with "Once the manager developed, ..."